File: chirp_control/lib/utils/scan_repo.dart
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import 'package:intl/intl.dart';
import 'package:csv/csv.dart';

class ScanData {
  final String folderName;
  final Directory folder;
  final List<List<dynamic>> sonarRows;
  final List<List<dynamic>> bathymetryRows;
  final String title;
  final String location;
  final String time;
  final String duration;

  ScanData({
    required this.folderName,
    required this.folder,
    required this.sonarRows,
    required this.bathymetryRows,
    required this.title,
    required this.location,
    required this.time,
    required this.duration,
  });
}

class ScanRepository {
  static double? _asDouble(dynamic value) {
    if (value is num) return value.toDouble();
    return double.tryParse(value.toString());
  }

  static String _formatDurationFromSeconds(int totalSeconds) {
    final minutes = totalSeconds ~/ 60;
    final seconds = totalSeconds % 60;
    return '${minutes}m ${seconds.toString().padLeft(2, '0')}s';
  }

  static DateTime? _extractFirstTimestamp(
    List<List<dynamic>> sonarRows,
    List<List<dynamic>> bathymetryRows,
  ) {
    if (bathymetryRows.isNotEmpty) {
      for (final row in bathymetryRows) {
        if (row.length > 4) {
          final ts = _asDouble(row[4]);
          if (ts != null) {
            return DateTime.fromMillisecondsSinceEpoch(ts.round());
          }
        }
      }
    }

    if (sonarRows.isNotEmpty) {
      for (final row in sonarRows) {
        if (row.isNotEmpty) {
          final ts = _asDouble(row[0]);
          if (ts != null && ts > 1000000000000) {
            return DateTime.fromMillisecondsSinceEpoch(ts.round());
          }
        }
      }
    }

    return null;
  }

  static int _extractDurationSeconds(
    List<List<dynamic>> sonarRows,
    List<List<dynamic>> bathymetryRows,
  ) {
    double? firstTs;
    double? lastTs;

    if (bathymetryRows.isNotEmpty) {
      for (final row in bathymetryRows) {
        if (row.length > 4) {
          final ts = _asDouble(row[4]);
          if (ts == null) continue;

          firstTs ??= ts;
          lastTs = ts;
        }
      }
    } else if (sonarRows.isNotEmpty) {
      for (final row in sonarRows) {
        if (row.isEmpty) continue;

        final ts = _asDouble(row[0]);
        if (ts == null || ts < 1000000000000) continue;

        firstTs ??= ts;
        lastTs = ts;
      }
    }

    if (firstTs == null || lastTs == null) return 0;

    final seconds = ((lastTs - firstTs) / 1000).round();
    return seconds < 0 ? 0 : seconds;
  }

  static Future<List<ScanData>> loadScans() async {
    final appDir = await getApplicationDocumentsDirectory();
    final scansDir = Directory('${appDir.path}/scans');

    if (!await scansDir.exists()) {
      await scansDir.create(recursive: true);
      return [];
    }

    final folders =
        scansDir.listSync(recursive: false).whereType<Directory>().toList()
          ..sort((a, b) => a.path.compareTo(b.path));

    final scans = <ScanData>[];

    for (final folder in folders) {
      final folderName = folder.path.split(Platform.pathSeparator).last;

      final sonarFile = File('${folder.path}/sonar.csv');
      final bathyFile = File('${folder.path}/bathymetry.csv');

      List<List<dynamic>> sonarRows = [];
      List<List<dynamic>> bathymetryRows = [];

      if (await sonarFile.exists()) {
        final content = await sonarFile.readAsString();
        sonarRows = CsvToListConverter().convert(content, eol: '\n');
      }

      if (await bathyFile.exists()) {
        final content = await bathyFile.readAsString();
        bathymetryRows = CsvToListConverter().convert(content, eol: '\n');
      }

      final firstTimestamp = _extractFirstTimestamp(sonarRows, bathymetryRows);
      final durationSeconds = _extractDurationSeconds(
        sonarRows,
        bathymetryRows,
      );

      final formattedTime = firstTimestamp != null
          ? DateFormat('M/d/yyyy, h:mm a').format(firstTimestamp)
          : 'Unknown time';

      final formattedDuration = _formatDurationFromSeconds(durationSeconds);

      scans.add(
        ScanData(
          folderName: folderName,
          folder: folder,
          sonarRows: sonarRows,
          bathymetryRows: bathymetryRows,
          title: folderName.replaceAll('_', ' '),
          location: 'SITE A',
          time: formattedTime,
          duration: formattedDuration,
        ),
      );
    }

    return scans;
  }

  static Future<void> deleteScan(ScanData scan) async {
    if (await scan.folder.exists()) {
      await scan.folder.delete(recursive: true);
    }
  }
}

File: chirp-control-web/src/utils/scanRepo.ts
import { createStore, get, set, del, keys } from "idb-keyval";

// Browsers have no equivalent to the Flutter app's per-scan folder on the
// local filesystem, so each scan is a single IndexedDB record instead. The
// raw CSV text is kept (not pre-parsed) so re-parsing on load matches the
// original app's behavior of reading straight from the CSV files each time.
const scansStore = createStore("chirp-control-scans", "scans");

export type CsvCell = string | number;
export type CsvRow = CsvCell[];

export function parseCsv(text: string): CsvRow[] {
  return text
    .split(/\r?\n/)
    .filter((line) => line.trim().length > 0)
    .map((line) =>
      line.split(",").map((cell) => {
        const trimmed = cell.trim();
        if (trimmed === "") return trimmed;
        const num = Number(trimmed);
        return Number.isNaN(num) ? trimmed : num;
      }),
    );
}

interface StoredScan {
  id: string;
  folderName: string;
  sonarCsv: string;
  bathymetryCsv: string;
  title: string;
  location: string;
  notes: string;
}

export interface ScanData {
  id: string;
  folderName: string;
  sonarRows: CsvRow[];
  bathymetryRows: CsvRow[];
  title: string;
  location: string;
  time: string;
  duration: string;
  notes: string;
}

function asNumber(value: CsvCell | undefined): number | null {
  if (value === undefined) return null;
  if (typeof value === "number") return value;
  const parsed = Number(value);
  return Number.isNaN(parsed) ? null : parsed;
}

function extractFirstTimestamp(
  sonarRows: CsvRow[],
  bathymetryRows: CsvRow[],
): number | null {
  if (bathymetryRows.length > 0) {
    for (const row of bathymetryRows) {
      if (row.length > 4) {
        const ts = asNumber(row[4]);
        if (ts !== null) return Math.round(ts);
      }
    }
  }

  if (sonarRows.length > 0) {
    for (const row of sonarRows) {
      if (row.length > 0) {
        const ts = asNumber(row[0]);
        if (ts !== null && ts > 1_000_000_000_000) return Math.round(ts);
      }
    }
  }

  return null;
}

function extractDurationSeconds(
  sonarRows: CsvRow[],
  bathymetryRows: CsvRow[],
): number {
  let firstTs: number | null = null;
  let lastTs: number | null = null;

  if (bathymetryRows.length > 0) {
    for (const row of bathymetryRows) {
      if (row.length > 4) {
        const ts = asNumber(row[4]);
        if (ts === null) continue;
        firstTs ??= ts;
        lastTs = ts;
      }
    }
  } else if (sonarRows.length > 0) {
    for (const row of sonarRows) {
      if (row.length === 0) continue;
      const ts = asNumber(row[0]);
      if (ts === null || ts < 1_000_000_000_000) continue;
      firstTs ??= ts;
      lastTs = ts;
    }
  }

  if (firstTs === null || lastTs === null) return 0;

  const seconds = Math.round((lastTs - firstTs) / 1000);
  return seconds < 0 ? 0 : seconds;
}

function formatDurationFromSeconds(totalSeconds: number): string {
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  return `${minutes}m ${String(seconds).padStart(2, "0")}s`;
}

function formatTime(timestampMs: number): string {
  const date = new Date(timestampMs);
  const datePart = new Intl.DateTimeFormat("en-US", {
    month: "numeric",
    day: "numeric",
    year: "numeric",
  }).format(date);
  const timePart = new Intl.DateTimeFormat("en-US", {
    hour: "numeric",
    minute: "2-digit",
    hour12: true,
  }).format(date);
  return `${datePart}, ${timePart}`;
}

function toScanData(stored: StoredScan): ScanData {
  const sonarRows = stored.sonarCsv ? parseCsv(stored.sonarCsv) : [];
  const bathymetryRows = stored.bathymetryCsv
    ? parseCsv(stored.bathymetryCsv)
    : [];

  const firstTimestamp = extractFirstTimestamp(sonarRows, bathymetryRows);
  const durationSeconds = extractDurationSeconds(sonarRows, bathymetryRows);

  return {
    id: stored.id,
    folderName: stored.folderName,
    sonarRows,
    bathymetryRows,
    title: stored.title.trim() || stored.folderName.replaceAll("_", " "),
    location: stored.location.trim() || "SITE A",
    time: firstTimestamp !== null ? formatTime(firstTimestamp) : "Unknown time",
    duration: formatDurationFromSeconds(durationSeconds),
    notes: stored.notes,
  };
}

export async function loadScans(): Promise<ScanData[]> {
  const allKeys = await keys(scansStore);
  const stored = await Promise.all(
    allKeys.map((key) => get<StoredScan>(key, scansStore)),
  );
  return stored
    .filter((s): s is StoredScan => !!s)
    .map(toScanData)
    .sort((a, b) => a.folderName.localeCompare(b.folderName));
}

export async function addScan(params: {
  folderName: string;
  sonarCsv: string;
  bathymetryCsv: string;
}): Promise<ScanData> {
  const stored: StoredScan = {
    id: crypto.randomUUID(),
    folderName: params.folderName,
    sonarCsv: params.sonarCsv,
    bathymetryCsv: params.bathymetryCsv,
    title: "",
    location: "",
    notes: "",
  };
  await set(stored.id, stored, scansStore);
  return toScanData(stored);
}

export async function renameScan(
  scan: ScanData,
  newTitle: string,
): Promise<void> {
  const trimmed = newTitle.trim();
  if (!trimmed) return;

  const stored = await get<StoredScan>(scan.id, scansStore);
  if (!stored) return;
  await set(scan.id, { ...stored, title: trimmed }, scansStore);
}

export async function saveNotes(scan: ScanData, notes: string): Promise<void> {
  const stored = await get<StoredScan>(scan.id, scansStore);
  if (!stored) return;
  await set(scan.id, { ...stored, notes }, scansStore);
}

export async function deleteScan(scan: ScanData): Promise<void> {
  await del(scan.id, scansStore);
}
